ERR: Improve error message on non-sorted input with .truncate

reidy-p opened the pull request:
- [ ] closes #17935
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

reidy-p committed:
ERR: Improve error message on non-sorted input with .truncate

jreback commented on pandas/tests/frame/test_timeseries.py:
```diff
@@ -374,6 +374,31 @@ def test_truncate_copy(self):
         truncated.values[:] = 5.
         assert not (self.tsframe.values[5:11] == 5).any()
 
+    def test_truncate_nonsortedindex(self):
```
can you add a test for series as well (you can also put this test generically in test_generic if its easier)

jreback reviewed (changes requested):
can you add a note in other api changes in 0.22 (this is technially an api change, sort of a trivial one)

jreback commented on pandas/core/generic.py:
```diff
@@ -6397,6 +6397,10 @@ def truncate(self, before=None, after=None, axis=None, copy=True):
         axis = self._get_axis_number(axis)
         ax = self._get_axis(axis)
 
```
add a comment here

reidy-p committed:
add tests and comments

jreback commented:
can you rebase

reidy-p committed:
clean whatsnewo

jreback commented on pandas/core/generic.py:
```diff
@@ -6338,6 +6338,12 @@ def truncate(self, before=None, after=None, axis=None, copy=True):
+        # GH 17935
+        # Check that index is sorted
+        if (not ax.is_monotonic_increasing and
+           not ax.is_monotonic_decreasing):
```
lint error

jreback reviewed (changes requested):
looks good, except you have lint issues, you can run ``make lint-diff`` and see

jreback commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -46,6 +46,8 @@ Other API Changes
 - :class:`Timestamp` will no longer silently ignore invalid ``freq`` arguments (:issue:`5168`)
 - :class:`CacheableOffset` and :class:`WeekDay` are no longer available in the ``pandas.tseries.offsets`` module (:issue:`17830`)
 - `tseries.frequencies.get_freq_group()` and `tseries.frequencies.DAYS` are removed from the public API (:issue:`18034`)
+- :func:`Series.truncate` and :func:`DataFrame.truncate` will raise a ``ValueError`` if the index is not sorted (:issue:`17935`)
```
say, rather than a non-helpful ``KeyError``

reidy-p committed:
fix whatsnew and failing test_truncate_nonsortedindex

reidy-p committed:
fix failing test_truncate

reidy-p commented:
I updated the whatsnew and think I fixed the lint issue.

But since I opened this PR a new the ``test_truncate`` test has been added to ``pandas/tests/series/test_period.py`` in #17755. This new test fails on my branch because it tries to truncate on a non-sorted index. So I modified the test to sort the index before truncating and now it passes. Alternatively, I could simply check for the new error message ``ValueError: truncate requires a sorted index`` in this test.

reidy-p committed:
more lint issues

jreback commented:
thanks @reidy-p 

changing the test for Period was just fine
